Testing new Docker buildx got me Java error on build

I have just tested the new cross-build features of Docker buildx command but found a problem reported here: https://github.com/docker/for-mac/issues/3646

I would like to test this with a newer Qemu version since Linuxkit uses 2.8.1 and the latest version is 4.0.

`qemu-aarch64 version 2.8.1(Debian 1:2.8+dfsg-6+deb9u4)`

What's the best way to test this?

I successfully built the VM using `linuxkit build -format iso-efi docker-for-mac.yml`:

```
Extract kernel image: linuxkit/kernel:4.9.125-4ffac525e6a57ccc3f2a8ae0fb96f12169027759-amd64
Add init containers:
Process init image: linuxkit/vpnkit-expose-port:v0.7
Process init image: linuxkit/init:v0.7
Process init image: linuxkit/runc:v0.7
Process init image: linuxkit/containerd:v0.7
Process init image: linuxkit/ca-certificates:v0.7
Add onboot containers:
  Create OCI config for linuxkit/metadata:v0.7
  Create OCI config for linuxkit/sysctl:v0.7
  Create OCI config for linuxkit/sysfs:v0.7
  Create OCI config for carlosedp/binfmt:1b9e0a3a0a6bd07e7875581f0e4498ea561dde10-dirty
  Create OCI config for linuxkit/format:v0.7
  Create OCI config for linuxkit/mount:v0.7
  Create OCI config for linuxkit/swap:v0.7
  Create OCI config for alpine:3.9
  Create OCI config for alpine:3.9
  Create OCI config for linuxkit/dhcpcd:v0.7
Add service containers:
  Create OCI config for linuxkit/acpid:v0.7
  Create OCI config for linuxkit/getty:v0.7
  Create OCI config for linuxkit/openntpd:v0.7
  Create OCI config for linuxkit/vsudd:v0.7
  Create OCI config for linuxkit/vpnkit-forwarder:v0.7
  Create OCI config for linuxkit/trim-after-delete:v0.7
  Create OCI config for linuxkit/host-timesync-daemon:v0.7
  Create OCI config for docker:19.03.0-beta3-dind
Create outputs:
  docker-for-mac-efi.iso
```

Started it with `linuxkit run hyperkit -networking=vpnkit -vsock-ports=2376 -disk size=4096M -data-file ./metadata.json -iso -uefi docker-for-mac-efi.iso`.

Apparently it booted but I get lots of `vpnkit: [ERROR] Failed to forward unexpected ICMP datagram with type 5`, then I pointed my Docker client to it and got:

```
➜ docker -H unix://docker-for-mac-efi-state/guest.00000948 version
Client: Docker Engine - Community
 Version:           19.03.0-beta3
 API version:       1.40
 Go version:        go1.12.4
 Git commit:        c55e026
 Built:             Thu Apr 25 19:05:38 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.0-beta3
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.4
  Git commit:       c55e026
  Built:            Thu Apr 25 03:02:31 2019
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          v1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc7+dev
  GitCommit:        029124da7af7360afa781a0234d1b083550f797c
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

But when trying to run any image:

```
➜ docker -H unix://docker-for-mac-efi-state/guest.00000948 run --rm hello-world
docker: Error response from daemon: OCI runtime create failed: container_linux.go:345: starting container process caused "process_linux.go:430: container init caused \"process_linux.go:413: running prestart hook 0 caused \\\"fork/exec /proc/6/exe: no such file or directory\\\"\"": unknown.
```

The changes I did to the yaml are using my binfmt image and updating to beta docker:

```yaml
# This is an example for building the open source components of Docker for Mac
kernel:
  image: linuxkit/kernel:4.9.125-4ffac525e6a57ccc3f2a8ae0fb96f12169027759-amd64
  cmdline: "console=ttyS0 page_poison=1"
init:
  - linuxkit/vpnkit-expose-port:v0.7 # install vpnkit-expose-port and vpnkit-iptables-wrapper on host
  - linuxkit/init:v0.7
  - linuxkit/runc:v0.7
  - linuxkit/containerd:v0.7
  - linuxkit/ca-certificates:v0.7
onboot:
  # support metadata for optional config in /run/config
  - name: metadata
    image: linuxkit/metadata:v0.7
  - name: sysctl
    image: linuxkit/sysctl:v0.7
  - name: sysfs
    image: linuxkit/sysfs:v0.7
  - name: binfmt
    image: carlosedp/binfmt:1b9e0a3a0a6bd07e7875581f0e4498ea561dde10-dirty
  # Format and mount the disk image in /var/lib/docker
  - name: format
    image: linuxkit/format:v0.7
  - name: mount
    image: linuxkit/mount:v0.7
    command: ["/usr/bin/mountie", "/var/lib"]
  # make a swap file on the mounted disk
  - name: swap
    image: linuxkit/swap:v0.7
    command: ["/swap.sh", "--path", "/var/lib/swap", "--size", "1024M"]
  # mount-vpnkit mounts the 9p share used by vpnkit to coordinate port forwarding
  - name: mount-vpnkit
    image: alpine:3.9
    binds:
        - /var/:/host_var:rbind,rshared
    capabilities:
        - CAP_SYS_ADMIN
    rootfsPropagation: shared
    command: ["sh", "-c", "mkdir -p /host_var/vpnkit/port && mount -v -t 9p -o trans=virtio,dfltuid=1001,dfltgid=50,version=9p2000 port /host_var/vpnkit"]
  # move logs to the mounted disk (this is a temporary fix until we can limit the log sizes)
  - name: move-logs
    image: alpine:3.9
    binds:
        - /var:/host_var
    command: ["sh", "-c", "mv -v /host_var/log /host_var/lib && ln -vs /var/lib/log /host_var/log"]
  - name: dhcpcd
    image: linuxkit/dhcpcd:v0.7
    command: ["/sbin/dhcpcd", "--nobackground", "-f", "/dhcpcd.conf", "-1"]
services:
  # Enable acpi to shutdown on power events
  - name: acpid
    image: linuxkit/acpid:v0.7
  # Enable getty for easier debugging
  - name: getty
    image: linuxkit/getty:v0.7
    env:
        - INSECURE=true
  # Run ntpd to keep time synchronised in the VM
  - name: ntpd
    image: linuxkit/openntpd:v0.7
  # VSOCK to unix domain socket forwarding. Forwards guest /var/run/docker.sock
  # to a socket on the host.
  - name: vsudd
    image: linuxkit/vsudd:v0.7
    binds:
        - /var/run:/var/run
    command: ["/vsudd", "-inport", "2376:unix:/var/run/docker.sock"]
  # vpnkit-forwarder forwards network traffic to/from the host via VSOCK port 62373. 
  # It needs access to the vpnkit 9P coordination share 
  - name: vpnkit-forwarder
    image: linuxkit/vpnkit-forwarder:v0.7
    binds:
        - /var/vpnkit:/port
    net: host
    command: ["/vpnkit-forwarder", "-vsockPort", "62373"]
  # Monitor for image deletes and invoke a TRIM on the container filesystem
  - name: trim-after-delete
    image: linuxkit/trim-after-delete:v0.7
  # When the host resumes from sleep, force a clock resync
  - name: host-timesync-daemon
    image: linuxkit/host-timesync-daemon:v0.7
  # Run dockerd with the vpnkit userland proxy from the vpnkit-forwarder container.
  # Bind mounts /var/run to allow vsudd to connect to docker.sock, /var/vpnkit
  # for vpnkit coordination and /run/config/docker for the configuration file.
  - name: docker-dfm
    image: docker:19.03.0-beta3-dind
    capabilities:
     - all
    net: host
    mounts:
     - type: cgroup
       options: ["rw","nosuid","noexec","nodev","relatime"]
    binds:
     - /etc/resolv.conf:/etc/resolv.conf
     - /var/lib/docker:/var/lib/docker
     - /lib/modules:/lib/modules
     - /var/vpnkit:/port # vpnkit control 9p mount
     - /var/run:/var/run
     - /run/config/docker:/var/config/docker
     - /usr/bin/vpnkit-expose-port:/usr/bin/vpnkit-expose-port # userland proxy
     - /usr/bin/vpnkit-iptables-wrapper:/usr/bin/iptables # iptables wrapper
    command: [ "/usr/local/bin/docker-init", "/usr/local/bin/dockerd", "--",
            "--config-file", "/var/config/docker/daemon.json",
            "--swarm-default-advertise-addr=eth0",
            "--userland-proxy-path", "/usr/bin/vpnkit-expose-port",
            "--storage-driver", "overlay2" ]
    runtime:
      mkdir: ["/var/lib/docker"]

# trust:
#     org:
#         - linuxkit
#         - library

```

--- a/examples/docker-for-mac.yml
+++ b/examples/docker-for-mac.yml
@@ -17,7 +17,7 @@ onboot:
   - name: sysfs
     image: linuxkit/sysfs:697f996868d6625a78bbbefb199d7b6b6ae3533a
   - name: binfmt
-    image: linuxkit/binfmt:94d59bc179c7a9155d8593a974d499352b70bfd5
+    image: linuxkit/binfmt:4ea3b9b0938cbd19834c096aa31ff475cc75d281
   # Format and mount the disk image in /var/lib/docker
   - name: format
     image: linuxkit/format:0b75e494eea0312f3015e6c6f7c5927620d56c96

--- a/pkg/binfmt/Dockerfile
+++ b/pkg/binfmt/Dockerfile
@@ -1,5 +1,5 @@
-# Use Debian stretch until https://bugs.alpinelinux.org/issues/8131 is resolved.
-FROM debian@sha256:de3eac83cd481c04c5d6c7344cd7327625a1d8b2540e82a8231b5675cef0ae5f AS qemu
+# Use Debian testing Qemu 4.2.0 until https://bugs.alpinelinux.org/issues/8131 is resolved.
+FROM debian@sha256:79f148e13b4c596d4ad7fd617aba3630e37cf04f04538699341ed1267bd61a23 AS qemu
 RUN apt-get update && apt-get install -y qemu-user-static && \
     mv /usr/bin/qemu-aarch64-static /usr/bin/qemu-aarch64 && \
     mv /usr/bin/qemu-arm-static /usr/bin/qemu-arm && \

--- a/test/cases/040_packages/002_binfmt/test.yml
+++ b/test/cases/040_packages/002_binfmt/test.yml
@@ -6,7 +6,7 @@ init:
   - linuxkit/runc:f79954950022fea76b8b6f10de58cb48e4fb3878
 onboot:
   - name: binfmt
-    image: linuxkit/binfmt:94d59bc179c7a9155d8593a974d499352b70bfd5
+    image: linuxkit/binfmt:4ea3b9b0938cbd19834c096aa31ff475cc75d281
   - name: test
     image: alpine:3.11
     binds:

